cpp/include/raft/sparse/solver/detail/lanczos.cuh

```diff
@@ -1396,4 +1436,682 @@ int computeLargestEigenvectors(
+      beta.data_handle(), 1, nEigVecs);
+    raft::matrix::fill(handle, beta_view, zero);
+
+    raft::copy(alpha.data_handle(), eigenvalues_k.data_handle(), nEigVecs, stream);
+
+    auto x_T =
+      raft::make_device_matrix_view<ValueTypeT>(ritz_eigenvectors.data_handle(), nEigVecs, n);
+
+    raft::copy(V.data_handle(), x_T.data_handle(), nEigVecs * n, stream);
+
+    ValueTypeT one  = 1;
+    ValueTypeT mone = -1;
+
+    // Using raft::linalg::gemv leads to Reason=7:CUBLAS_STATUS_INVALID_VALUE
+    raft::linalg::detail::cublasgemv(cublas_h,
+                                     CUBLAS_OP_T,
+                                     nEigVecs,
+                                     n,
+                                     &one,
+                                     V.data_handle(),
+                                     nEigVecs,
+                                     u.data_handle(),
+                                     1,
+                                     &zero,
+                                     uu.data_handle(),
+                                     1,
+                                     stream);
+
+    raft::linalg::detail::cublasgemv(cublas_h,
+                                     CUBLAS_OP_N,
+                                     nEigVecs,
+                                     n,
+                                     &mone,
+                                     V.data_handle(),
+                                     nEigVecs,
+                                     uu.data_handle(),
+                                     1,
+                                     &one,
+                                     u.data_handle(),
+                                     1,
+                                     stream);
+
+    auto V_0_view =
+      raft::make_device_matrix_view<ValueTypeT>(V.data_handle() + (nEigVecs * n), 1, n);
+
+    auto unrm  = raft::make_device_vector<ValueTypeT, uint32_t>(handle, 1);
+    auto input = raft::make_device_matrix_view<const ValueTypeT, uint32_t>(u.data_handle(), 1, n);
+    raft::linalg::norm(handle,
+                       input,
+                       unrm.view(),
+                       raft::linalg::L2Norm,
+                       raft::linalg::Apply::ALONG_ROWS,
+                       raft::sqrt_op());
+
+    auto u_vector_const = raft::make_device_vector_view<const ValueTypeT>(u.data_handle(), n);
+
+    raft::linalg::unary_op(
+      handle, u_vector_const, V_0_view, [device_scalar = unrm.data_handle()] __device__(auto y) {
+        return y / *device_scalar;
+      });
+
+    auto cusparse_h = resource::get_cusparse_handle(handle);
+    cusparseSpMatDescr_t cusparse_A;
+    //   input_config.a_indptr  = const_cast<IndexType*>(x_structure.get_indptr().data());
+    // input_config.a_indices = const_cast<IndexType*>(x_structure.get_indices().data());
+    // input_config.a_data    = const_cast<ElementType*>(x.get_elements().data());
+    raft::sparse::detail::cusparsecreatecsr(
```

Comment:
This too- we really should be abstracting this away so that we are passing our own sparse types into public APIs and those APIs in turn invoke the detail namespaces.